README
Mauve, Clonal Frame, Clonal Origin Analysis
===========================================
Sang Chul Choi <goshng@yahoo.co.kr>

See doc/Manual for the instruction.

See manuscript/README for files that are used in the manuscript.

See ChangeLog for the history of the source code update.


ChangeLog
Mauve, Clonal Frame, Clonal Origin Analysis
===========================================

History
-------
Jobs that were done are described for later reference.

.Meaning of arrows
-> means I finished the job.
---> means I need to do something about it.

To Do
~~~~~

.alpha1, alpha2, and tau would be implemented.

.Finish works in the directory +manuscript+.

Running jobs
~~~~~~~~~~~~
NA


Mon Apr 16 10:46:50 EDT 2012
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.To get 5003 version
----
mkdir -p writing/latex
mkdir -p writing/papers
cd writing/papers
svn co $SVNROOT/writing/trunk/papers/strep-recomb@5003
make
bash create-final-submission.sh
----

----
svn co $SVNROOT/writing/tags/strep-recomb-mbe-v1/papers/strep-recomb
svn co $SVNROOT/writing/tags/strep-recomb-mbe-v2
----

.Summary for revision
----
pl/extract-cornell-species-tree.pl 
----

Fri Mar 23 16:04:03 EDT 2012
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.The table with phage and nonphage.
----
test/mowgli/fam-frag2/famid-events-phage.txt
source("famid-events-phage.R")
> length(x$V1)
[1] 2660
> sum(x$V6=="phage")
[1] 194
> sum(x$V4[x$V6=="phage"]>0)
[1] 83
> sum(x$V6=="nonphage")
[1] 2466
> sum(x$V4[x$V6=="nonphage"]>0)
[1] 165
----

1.out -> list of genes with phages

.event counts for families:
----
./test/mowgli/fam-frag2/famid-events.txt
compgen:/usr/projects/strep/gainloss/data/fam-frag2/famid-events.txt
----
The columns are
1. famid
2. dup count
3. loss count
4. additive transfer count
5. replacing transfer count

.gene family membership is stored here:
----
./test/mowgli/fam-frag2/famid-gene.txt
compgen:/usr/projects/strep/gainloss/data/fam-frag2/famid-gene.txt
----


Tue Mar 13 15:39:21 EDT 2012
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

We found the following many imports. It appears that SDE and SPY receive comparable number of imports.
----
SDE1: 176
SDE2: 172
SDD : 546
SPY1: 163
SPY2: 158
Read 176 items
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
    1.0    22.0    67.0   133.8   130.2  2091.0 
   2.5%     50%   97.5% 
  5.375  67.000 645.500 
[1] 23551
Read 172 items
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
    2.0    27.0    64.0   102.4   139.0  1158.0 
 2.5%   50% 97.5% 
  5.0  64.0 329.3 
[1] 17620
Read 546 items
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
    1.0    34.0    70.5   185.1   150.8 15130.0 
   2.5%     50%   97.5% 
  6.625  70.500 920.750 
[1] 101067
Read 163 items
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   1.00   16.00   31.00   77.68   75.50 1343.00 
 2.5%   50% 97.5% 
  4.0  31.0 434.3 
[1] 12662
Read 158 items
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   3.00   18.00   35.50   84.98   99.25 1336.00 
   2.5%     50%   97.5% 
  6.775  35.500 482.425 
[1] 13427
----

We need to read a ClonalFrame output file to locate parts that start with 
#consevents, #poly, and #blocks. Refer to ClonalFrame's manual for details. I
just copy explanations of them here for my information. 

I will find consecutive import regions with high recombination probabilities. We
set an array of values of recombination probabilities. The number of elements
in the array is equal to the size of SDE1's genome. Values at the sites referred
to by #poly are used. Because we wish to know imports only, we ignore sites not
specified in #poly. So, we delete those elements from the array. Now, we use a
threshold, say 0.9, to code each element being 0 or 1. We need to find regions
with consecutive 1's. Let's have a utility menu: summarize-clonalframe-event.

.The main ClonalFrame output
----
output/cornellf/3/run-clonalframe/output/1/core_clonalframe.out.1
----

.blocks
-----
It contains a list of numbers indicating the beginning and end of blocks of
alignment, in number of reference sites. For example the vector 2,3,2 indicates
that the first two reference belong to a block, the next three to another block
and the last two to a third block.
----

.poly
----
This part starts with the line #poly. It contains a vector indicating the
location of the reference sites on the first (reference) isolate. The reference
sites are the sites at the beginning or end of each block, the polymorphic sites
and all sites that are at a distance from the start of the block they belong to
which is a multiple of 50.
----

.consevents
----
This part starts with the line #consevents. It contains the events found on the
consensus tree of Section 7.1. It is made of a N × (2M) matrix where the number
of lines N is the number of nodes in the consensus tree, and the number of
columns is two times M where M is the number of reference sites (cf. Section
7.7). The first number of each pair for each reference site indicates the
posterior probability of recombination at that site and the second one the
posterior probability of substitution (either through recombination or through
mutation).
----

Tue Feb 14 13:36:01 EST 2012
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.Copy warg output files
----
cd an-empty-directory
perl ~/cac.pl -copywarg
----

Fri Feb 10 11:43:40 EST 2012
~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cornell1: MGAS10270, MGAS8232
cornell2: MGAS9429, MGAS10394
cornell3: MGAS6180, M1
cornell4: Alab49, NZ131
cornellfx5: Manfredo, 10394
cornellfx6: 2096, 6180
cornellfx7: 315, Alab49
cornellfx8: 10394, 8232
cornellfx9: 6180, 8232
cornellfx10: manfredo, 2096
I might want to change cornell1,2,3, and 4 to cornellfx1,2,3, and 4.

Data for pairs of genomes
swiftgen:/usr/projects/strep/gainloss/recombination/matt/genomes


Fri Feb  3 08:53:14 EST 2012
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Note: repetition 3 is used not 1.
cornell1, cornell2, cornell3: pairs of S. pyogenes replaced.
cornell4: cornellf + 6 additional S. pyogenes
pyogenes: 14 S. pyogenes
pyogenes11: 11 S. pyogenes
cornellf8: cornellf + 6 additional S. pyogenes, the 8 S. pyogenes are formed by
removing 3 (same serotypes), and 3 (wandering taxa).
Same types: MGAS5005, SSI_1, and MGAS9429
Wandering taxa: MGAS10270, NZ131, and M1GAS

Thu Jan 26 13:03:59 EST 2012
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

pl/cac.pl
#!/usr/bin/perl -w
use strict;
use warnings;

use File::Path qw(mkpath);
use Getopt::Long;
use Pod::Usage;

$| = 1; # Do not buffer output

my $VERSION = 'cac.pl 1.0';

my $man = 0;
my $help = 0;
my %params = ('help' => \$help, 'h' => \$help);    
GetOptions( \%params,
            'help|h',
            'verbose',
            'version' => sub { print $VERSION."\n"; exit; },
            'copywarg',
            'printjob',
            'printtmp',
            'printwarg',
            'killwarg',
            'jobidfile=s',
            'only=i',
            'command=s'
            ) or pod2usage(2);
pod2usage(1) if $help;
pod2usage(-exitstatus => 0, -verbose => 2) if $man;

=head1 NAME

cac.pl - monitor or kill warg jobs in compute nodes

=head1 VERSION

cac.pl 1.0

=head1 SYNOPSIS

perl cac.pl [-h] [-help] [-version] [-verbose] 

=head1 DESCRIPTION

perl cac.pl -printjob -command "ps -ef | grep warg | wc"

perl cac.pl -printjob

perl cac.pl -printwarg

perl cac.pl -killwarg

perl cac.pl -jobidfile y

perl cac.pl -printtmp

perl cac.pl -copywarg

=head1 OPTIONS

=over 8

=item B<-help> | B<-h>

Print the help message; ignore other arguments.

=item B<-version>

Print program version; ignore other arguments.

=item B<-verbose>

Prints status and info messages during processing.

=item B<***** COMMAND OPTIONS *****>

=item B<-copywarg>

perl cac.pl -copywarg

This allows to copy all of the output files of warg command. CO1's output files
are in directory output, and CO2's output files in directory output2. Because it
copy files to a directory named output. You can run the command at an empty
directory or the main output directory such as Documents/Projects/m2.

=back

=head1 AUTHOR

Sang Chul Choi, C<< <goshng_at_yahoo_dot_co_dot_kr> >>

=head1 BUGS

If you find a bug please post a message rnaseq_analysis project at codaset dot
com repository so that I can make compute-block-length.pl better.

=head1 COPYRIGHT

Copyright (C) 2011-2012 Sang Chul Choi

=head1 LICENSE

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.

=cut


my $printjob = 0;
my $killwarg = 0;
my $printwarg = 0;
my $printtmp = 0;
my $jobidfile; 
my $only;
my $command = "ps -ef | grep warg | wc -l";

if (exists $params{printtmp}) {
  $printtmp = 1;
} 

if (exists $params{printwarg}) {
  $printwarg = 1;
} 

if (exists $params{printjob}) {
  $printjob = 1;
} 

if (exists $params{killwarg}) {
  $killwarg = 1;
} 

if (exists $params{command}) {
  $command = $params{command};
} 

if (exists $params{only}) {
  $only = $params{only};
} 

if (exists $params{jobidfile}) {
  $jobidfile = $params{jobidfile};
} 


sub getComputeNode ();
sub getComputeJobID ();
sub printJob ($$);
sub printWarg ($$);
sub copyWarg ($$);
sub printTmp ($$);
sub killWarg ($);
sub jobIDFile ($);
sub jobIDFilexxx ();

if ($printjob == 1)
{
  my @n = getComputeNode ();
  printJob (\@n, $command);
} 
elsif ($printwarg == 1)
{
  my @n = getComputeNode ();
  my @jobid = getComputeJobID ();
  printWarg (\@n, \@jobid);    
}
elsif ($killwarg == 1)
{
  my @n = getComputeNode ();
  killWarg (\@n);    
}
elsif (defined ($jobidfile))
{
  my @n = getComputeNode ();
  jobIDFile (\@n);
}
elsif ($printtmp == 1)
{
  my @n = getComputeNode ();
  my @jobid = getComputeJobID ();
  printTmp (\@n, \@jobid);    
}
elsif (exists $params{copywarg}) 
{
  my @n = getComputeNode ();
  my @jobid = getComputeJobID ();
  copyWarg (\@n, \@jobid);    
}


sub jobIDFile ($)
{
  my ($address) = @_;
  open JOBIDFILE, ">$jobidfile";
  for (my $i = 0; $i <= $#{ $address }; $i++)
  {
    open WARG, "ssh -x $address->[$i] ps -ef | grep warg |";
    while (<WARG>)
    {
      #          ^^^^^  
      # sc2265   12887 12218 99 May12 ?        23:39:19 ./warg -a 1 1 0.1 1 1 1 1 1 0 0 0 -x 1000000 -y 1000000 -z 10000 input/2/cornellf-3.tree input/2/core_alignment.1.xmfa.210 output/2/1/core_co.phase2.xml.210
      if (/(\S+)\s+\.\/warg -a 1 1 0.1 1 1 1 1 1 0 0 0\s+-x\s+\d+\s+-y\s+\d+\s+-z\s+\d+\s+\S+\s+(\S+)\s+(\S+)/)
      {
        my $xml = $3;
        $xml =~ /output\/(\d+)\/(\d+)\/core_co.phase2.xml.(\d+)/;
        my $REPETITION = $1;
        my $REPLICATE = $2;
        my $BLOCK = $3;
        print JOBIDFILE "-a 1,1,0.1,1,1,1,1,1,0,0,0 -x 1000000 -y 1000000 -z 10000 input/$REPETITION/cornellf-3.tree input/$REPETITION/core_alignment.$REPLICATE.xmfa.$BLOCK output/$REPETITION/$REPLICATE/core_co.phase2.xml.$BLOCK\n";
      }
    }
    close WARG;
  }
  close JOBIDFILE;
}

sub jobIDFilexxx ()
{
  open JOBIDFILE, ">$jobidfile";

  open X, "x";
  while (<X>)
  {
    # output/7/1/core_co.phase2.xml.113 (17:02:04):      55      55    1375
    if (/output\/(\d+)\/(\d+)\/core_co.phase2.xml.(\d+)/)
    {
      my $REPETITION = $1;
      my $REPLICATE = $2;
      my $BLOCK = $3;
      print JOBIDFILE "-a 1,1,0.1,1,1,1,1,1,0,0,0 -x 1000000 -y 1000000 -z 10000 input/$REPETITION/cornellf-3.tree input/$REPETITION/core_alignment.$REPLICATE.xmfa.$BLOCK output/$REPETITION/$REPLICATE/core_co.phase2.xml.$BLOCK\n";
    }
  }
  close X;
  
  close JOBIDFILE;
}

sub killWarg ($)
{
  my ($address) = @_;
  for (my $i = 0; $i <= $#{ $address }; $i++)
  {
    open WARG, "ssh -x $address->[$i] ps -ef | grep warg |";
    while (<WARG>)
    {
      #          ^^^^^  
      # sc2265   12887 12218 99 May12 ?        23:39:19 ./warg -a 1 1 0.1 1 1 1 1 1 0 0 0 -x 1000000 -y 1000000 -z 10000 input/2/cornellf-3.tree input/2/core_alignment.1.xmfa.210 output/2/1/core_co.phase2.xml.210
      if (/(\w+)\s+(\d+)/)
      {
        #print "ssh -x $address->[$i] kill $2\n";
        system "ssh -x $address->[$i] kill $2";
      }
    }
    close WARG;
  }
}

sub printTmp ($$)
{
  my ($address, $jobid) = @_;
  for (my $i = 0; $i <= $#{ $address }; $i++)
  {

    print "ssh -x $address->[$i] ls /tmp/$jobid->[$i]/\n";
  }
}


sub printWarg ($$)
{
  my ($address, $jobid) = @_;
  for (my $i = 0; $i <= $#{ $address }; $i++)
  {
    print "$address->[$i]\n";
    open WARG, "ssh -x $address->[$i] ps -ef | grep warg |";
    while (<WARG>)
    {
      # sc2265   12887 12218 99 May12 ?        23:39:19 ./warg -a 1 1 0.1 1 1 1 1 1 0 0 0 -x 1000000 -y 1000000 -z 10000 input/2/cornellf-3.tree input/2/core_alignment.1.xmfa.210 output/2/1/core_co.phase2.xml.210
      #print $_, "\n";
      if (/(\S+)\s+\.\/warg -a 1 1 0.1 1 1 1 1 1 0 0 0\s+-x\s+\d+\s+-y\s+\d+\s+-z\s+\d+\s+\S+\s+(\S+)\s+(\S+)/)
      {
        if ($3 eq "-D")
        {
# sc2265    4586  4585 95 21:17 ?        00:10:55 ./warg -a 1 1 0.1 1 1 1 1 1 0 0
# 0 -x 10000000 -y 100000000 -z 100000 -T s0.080814517322956 -D 743.702415841584
# -R s0.0118491287894766 input/3/clonaltree.nwk input/3/core_alignment.xmfa.274
# output/3/core_co.phase3.xml.274
          if (/(\S+)\s+\.\/warg -a 1 1 0.1 1 1 1 1 1 0 0 0\s+-x\s+\d+\s+-y\s+\d+\s+-z\s+\d+\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+(\S+)\s+(\S+)/)
          {
            print "$3 ($1): ";
            system "ssh -x $address->[$i] grep number /tmp/$jobid->[$i]/$3 | wc -l";
            #my $id = $3;
            #$id =~ /\.xml\.(\d+)/;
            #print "$1\t";
          }
        }
        else
        {
          print "$3 ($1): ";
          system "ssh -x $address->[$i] grep number /tmp/$jobid->[$i]/$3 | wc -l";
          #my $id = $3;
          #$id =~ /\.xml\.(\d+)/;
          #print "$1\t";
        }
      }
    }
    close WARG;
  }
}

sub copyWarg ($$)
{
  my ($address, $jobid) = @_;
  for (my $i = 0; $i <= $#{ $address }; $i++)
  {
    open WARG, "ssh -x $address->[$i] ps -ef | grep warg |";
    while (<WARG>)
    {
      # sc2265   12887 12218 99 May12 ?        23:39:19 ./warg -a 1 1 0.1 1 1 1 1 1 0 0 0 -x 1000000 -y 1000000 -z 10000 input/2/cornellf-3.tree input/2/core_alignment.1.xmfa.210 output/2/1/core_co.phase2.xml.210
      #print $_, "\n";
      my $outputdir = "";
      if (/(\S+)\s+\.\/warg -a 1 1 0.1 1 1 1 1 1 0 0 0\s+-x\s+\d+\s+-y\s+\d+\s+-z\s+\d+\s+\S+\s+(\S+)\s+(\S+)/)
      {
        if ($3 eq "-D")
        {
          # CO2
# sc2265    4586  4585 95 21:17 ?        00:10:55 ./warg -a 1 1 0.1 1 1 1 1 1 0 0
# 0 -x 10000000 -y 100000000 -z 100000 -T s0.080814517322956 -D 743.702415841584
# -R s0.0118491287894766 input/3/clonaltree.nwk input/3/core_alignment.xmfa.274
# output/3/core_co.phase3.xml.274
          if (/(\S+)\s+\.\/warg -a 1 1 0.1 1 1 1 1 1 0 0 0\s+-x\s+\d+\s+-y\s+\d+\s+-z\s+\d+\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+\S+\s+(\S+)\s+(\S+)/)
          {
            #print "CO2 $3 ($1): ";
            my $outputfile = $3;
            $outputfile =~ /(.+)\//;
            $outputdir = $1;
            #unless (-d $outputdir) {
              #mkpath $outputdir;
            #} 
            #system "scp -q $address->[$i]:/tmp/$jobid->[$i]/$outputfile $outputfile";
          }
        }
        else
        {
          # CO1
          #print "CO1 $3 ($1): ";
          my $outputfile = $3;
          $outputfile =~ /(.+)\//;
          $outputdir = $1;
          #unless (-d $outputdir) {
            #mkpath $outputdir;
          #} 
          #system "scp -q $address->[$i]:/tmp/$jobid->[$i]/$outputfile $outputfile";
        }
      }

      # 
      my $outputbasedir;
      my $codir;
      if (length($outputdir) > 0)
      {
        $outputdir =~ /(.+)\//;
        $outputbasedir = $1;
        $outputbasedir =~ /(.+)\//;
        $codir = $1;
        unless (-d $outputbasedir) {
          mkpath $outputbasedir;
        } 
        system "scp -qr $address->[$i]:/tmp/$jobid->[$i]/$outputbasedir $codir";
        # print "\n";
        last;
      }
    }
    close WARG;
  }
}

sub printJob ($$)
{
  my ($address, $command) = @_;
  for (my $i = 0; $i <= $#{ $address }; $i++)
  {
    print "$address->[$i]\t";
    system ("ssh -x $address->[$i] $command");
  }
}

sub getComputeNode ()
{
  my @v;
  open QSTAT, "qstat -f|";
  while (<QSTAT>)
  {
    if (/(compute-\d+-\d+)\.v4linux/)
    {
      push @v, $1;
    }
  }
  close QSTAT;
  return @v; 
}

sub getComputeJobID ()
{
  my @v;
  open QSTAT, "qstat -f|";
  while (<QSTAT>)
  {
    if (/Job Id:\s+(.+)/)
    {
      push @v, $1;
    }
  }
  close QSTAT;
  return @v;
}
